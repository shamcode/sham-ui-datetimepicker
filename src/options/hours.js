import { ref } from 'sham-ui-macro/ref.macro';
import { formatDate, clone, change, compare } from '../utils/date';
import { classes } from '../utils/string';

export function hoursOptions( component ) {
    return {
        [ ref( 'showChange' ) ]: false,
        [ ref( 'header' ) ]( date ) {
            return formatDate( date, 1 + 2 + 4 + 8 );
        },
        [ ref( 'items' ) ]( current, date, classForDate, isDateSelectable ) {
            const hours = new Array( 24 );
            let d = new Date(
                date.getFullYear(),
                date.getMonth(),
                date.getDate(),
                0,
                date.getMinutes(),
                date.getSeconds()
            );
            for ( let i = 0; i < 24; i++ ) {
                const hour = clone( d );
                hours[ i ] = {
                    date: hour,
                    class: classes( [
                        classForDate( hour ),
                        compare( current, hour, 1 + 2 + 4 + 8 ) ? 'active' : '',
                        isDateSelectable( hour ) ? '' : 'disabled'
                    ] ),
                    label: formatDate( hour, 16 )
                };
                d = change( d, 'Hours', 1 );
            }
            return hours;
        },
        [ ref( 'clickOnHeader' ) ]() {
            component[ ref( 'changeMode' ) ]( 'dom' );
        },
        [ ref( 'select' ) ]( date ) {
            component[ ref( 'changeValue' ) ]( date );
            component[ ref( 'changeMode' ) ]( 'time' );
        }
    };
}
