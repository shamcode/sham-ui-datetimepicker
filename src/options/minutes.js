import { ref } from 'sham-ui-macro/ref.macro';
import { formatDate, change, clone } from '../utils/date';
import { classes } from '../utils/string';

export function minutesOptions( component ) {
    return {
        [ ref( 'showChange' ) ]: false,
        [ ref( 'header' ) ]( date ) {
            return formatDate( date, 1 + 2 + 4 + 8 );
        },
        [ ref( 'items' ) ]( current, date, classForDate, isDateSelectable ) {
            const minutes = new Array( 12 );
            let d = new Date(
                date.getFullYear(),
                date.getMonth(),
                date.getDate(),
                date.getHours(),
                0,
                date.getSeconds()
            );
            for ( let i = 0; i < 12; i++ ) {
                const minute = clone( d );
                minutes[ i ] = {
                    date: minute,
                    class: classes( [
                        classForDate( minute ),
                        isDateSelectable( minute ) ? '' : 'disabled'
                    ] ),
                    label: formatDate( minute, 32 )
                };
                d = change( d, 'Minutes', 5 );
            }
            return minutes;
        },
        [ ref( 'clickOnHeader' ) ]() {
            component[ ref( 'changeMode' ) ]( 'dom' );
        },
        select( date ) {
            component[ ref( 'changeValue' ) ]( date );
            component[ ref( 'changeMode' ) ]( 'time' );
        }
    };
}
