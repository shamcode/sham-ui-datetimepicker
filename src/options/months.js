import { ref } from 'sham-ui-macro/ref.macro';
import { title, classes } from '../utils/string';
import { clone, change, compare, formatDate } from '../utils/date';

export function monthOptions( component ) {
    return {
        [ ref( 'showChange' ) ]: true,
        [ ref( 'prevDelta' ) ]: -1,
        [ ref( 'nextDelta' ) ]: 1,
        [ ref( 'header' ) ]( date ) {
            return formatDate( date, 1 );
        },
        [ ref( 'items' ) ]( current, date, classForDate, isDateSelectable ) {
            const months = new Array( 12 );
            let d = new Date( date.getFullYear(), 0, 1 );
            for ( let i = 0; i < 12; i++ ) {
                const month = clone( d );
                months[ i ] = {
                    date: month,
                    class: classes( [
                        'month',
                        classForDate( month ),
                        compare( current, month, 1 + 2 ) ? 'active' : '',
                        isDateSelectable( month ) ? '' : 'disabled'
                    ] ),
                    label: title(
                        formatDate( month, 256 )
                    )
                };
                d = change( d, 'Month', 1 );
            }
            return months;
        },
        [ ref( 'clickOnHeader' ) ]() {
            component[ ref( 'changeMode' ) ]( 'year' );
        },
        change( delta ) {
            component[ ref( 'changeDate' ) ](
                change( component.options[ ref( 'date' ) ], 'FullYear', delta )
            );
        },
        select( date ) {
            component[ ref( 'changeDate' ) ]( date );
            component[ ref( 'changeMode' ) ]( 'dom' );
        }
    };
}
