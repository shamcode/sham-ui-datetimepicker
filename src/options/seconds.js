import { ref } from 'sham-ui-macro/ref.macro';
import { formatDate, change, clone } from '../utils/date';
import { classes } from '../utils/string';

export function secondsOptions( component ) {
    return {
        [ ref( 'showChange' ) ]: false,
        [ ref( 'header' ) ]( date ) {
            return formatDate( date, 1 + 2 + 4 + 8 );
        },
        [ ref( 'items' ) ]( current, date, classForDate, isDateSelectable ) {
            const seconds = new Array( 12 );
            let d = new Date(
                date.getFullYear(),
                date.getMonth(),
                date.getDate(),
                date.getHours(),
                date.getMinutes()
            );
            for ( let i = 0; i < 12; i++ ) {
                const second = clone( d );
                seconds [ i ] = {
                    date: second,
                    class: classes( [
                        classForDate( second ),
                        isDateSelectable( second ) ? '' : 'disabled'
                    ] ),
                    label: formatDate( second, 64 )
                };
                d = change( d, 'Seconds', 5 );
            }
            return seconds;
        },
        [ ref( 'clickOnHeader' ) ]() {
            component[ ref( 'changeMode' ) ]( 'dom' );
        },
        select( date ) {
            component[ ref( 'changeValue' ) ]( date );
            component[ ref( 'changeMode' ) ]( 'time' );
        }
    };
}
