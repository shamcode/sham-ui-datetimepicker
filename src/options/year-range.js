import { ref } from 'sham-ui-macro/ref.macro';
import { formatDate, change } from '../utils/date';
import { classes } from '../utils/string';

const getCentury = ( date ) => 100 * Math.floor( date.getFullYear() / 100 );

const isCurrentYearRange = ( current, yearRange ) => {
    const year = current.getFullYear();
    return yearRange.from <= year &&
        yearRange.to >= year;
};

export function yearsRangeOptions( component ) {
    return {
        [ ref( 'showChange' ) ]: true,
        [ ref( 'prevDelta' ) ]: -100,
        [ ref( 'nextDelta' ) ]: 100,
        [ ref( 'header' ) ]( date ) {
            const century = getCentury( date );
            const from = new Date( century - 1, 0, 1 );
            const to = new Date( century + 99, 0, 1 );
            return `${formatDate( from, 1 )} - ${formatDate( to, 1 )}`;
        },
        [ ref( 'items' ) ]( current, date ) {
            const years = new Array( 10 );
            const century = getCentury( date );
            let d = new Date( century, 0, 1 );
            for ( let i = 0; i < 10; i++ ) {
                const from  = d.getFullYear();
                const to = from + 9;
                const date = new Date( from + 5, 0, 1 );
                const range = {
                    from,
                    to,
                    date,
                    label: `${formatDate( new Date( from, 0, 1 ), 1 )} - ${formatDate( new Date( to, 0, 1 ), 1 )}`
                };
                range.class = classes( [
                    'decade',
                    isCurrentYearRange( current, range ) ? 'active' : ''
                ] );
                years[ i ] = range;
                d = change( d, 'FullYear', 10 );
            }
            return years;
        },
        [ ref( 'change' ) ]( delta ) {
            component[ ref( 'changeDate' ) ](
                change( component.options[ ref( 'date' ) ], 'FullYear', delta )
            );
        },
        [ ref( 'select' ) ]( date ) {
            component[ ref( 'changeDate' ) ]( date );
            component[ ref( 'changeMode' ) ]( 'year' );
        }
    };
}
