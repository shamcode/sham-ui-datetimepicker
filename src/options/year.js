import { ref } from 'sham-ui-macro/ref.macro';
import { formatDate, change, clone, compare } from '../utils/date';
import { classes } from '../utils/string';

export function yearOptions( component ) {
    return {
        [ ref( 'showChange' ) ]: true,
        [ ref( 'prevDelta' ) ]: -10,
        [ ref( 'nextDelta' ) ]: 10,
        [ ref( 'header' ) ]( date ) {
            const from = new Date( date.getFullYear() - 5, 0, 1 );
            const to = new Date( date.getFullYear() + 6, 0, 1 );
            return `${formatDate( from, 1 )} - ${formatDate( to, 1 )}`;
        },
        [ ref( 'items' ) ]( current, date, classForDate, isDateSelectable ) {
            const years = new Array( 12 );
            let d = new Date( date.getFullYear() - 5, 0, 1 );
            for ( let i = 0; i < 12; i++ ) {
                const year = clone( d );
                years[ i ] = {
                    date: year,
                    class: classes( [
                        'year',
                        classForDate( year ),
                        compare( current, year, 1 ) ? 'active' : '',
                        isDateSelectable( year ) ? '' : 'disabled'
                    ] ),
                    label: formatDate( year, 1 )
                };
                d = change( d, 'FullYear', 1 );
            }
            return years;
        },
        [ ref( 'clickOnHeader' ) ]() {
            component[ ref( 'changeMode' ) ]( 'yearRange' );
        },
        change( delta ) {
            component[ ref( 'changeDate' ) ](
                change( component.options[ ref( 'date' ) ], 'FullYear', delta )
            );
        },
        select( date ) {
            component[ ref( 'changeDate' ) ]( date );
            component[ ref( 'changeMode' ) ]( 'months' );
        }
    };
}
