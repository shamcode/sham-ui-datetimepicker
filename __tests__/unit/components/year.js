import ItemsList  from '../../../src/components/item-list.sfc';
import { yearOptions } from '../../../src/options/year';
import renderer from 'sham-ui-test-helpers';
import { onclick } from 'sham-ui-directives';

it( 'renders correctly', () => {
    const meta = renderer( ItemsList, {
        today: new Date( 2020, 3, 25, 15, 3, 30 ),
        value: new Date( 2020, 3, 23, 13, 2, 17 ),
        date: new Date( 2020, 3, 23, 13, 2, 17 ),
        directives: {
            onclick
        },
        ...yearOptions()
    } );
    expect( meta.toJSON() ).toMatchSnapshot();
} );


it( 'change', () => {
    const changeDate = jest.fn();
    const changeMode = jest.fn();
    const value =  new Date( 2020, 3, 23, 13, 2, 17 );
    const meta = renderer( ItemsList, {
        today: new Date( 2020, 3, 25, 15, 3, 30 ),
        value,
        date: new Date( 2020, 3, 23, 13, 2, 17 ),
        directives: {
            onclick
        },
        isDateSelectable( date ) {
            return date.getFullYear() <= 2020;
        },
        ...yearOptions( {
            changeMode,
            changeDate,
            options: {
                date: new Date( value.getTime() )
            }
        } )
    } );
    expect( meta.toJSON() ).toMatchSnapshot();

    meta.component.container.querySelector( 'thead tr th:first-child' ).click();
    expect( changeDate ).toHaveBeenCalledTimes( 1 );
    expect( changeDate.mock.calls[ 0 ][ 0 ].toISOString() ).toBe( '2010-04-23T06:02:17.000Z' );
    expect( meta.toJSON() ).toMatchSnapshot();

    meta.component.container.querySelector( 'thead tr th:last-child' ).click();
    expect( changeDate ).toHaveBeenCalledTimes( 2 );
    expect( changeDate.mock.calls[ 1 ][ 0 ].toISOString() ).toBe( '2020-04-23T06:02:17.000Z' );
    expect( meta.toJSON() ).toMatchSnapshot();

    meta.component.container.querySelector( 'tbody span:nth-child(3)' ).click();
    expect( changeDate ).toHaveBeenCalledTimes( 3 );
    expect( changeDate.mock.calls[ 2 ][ 0 ].toISOString() ).toBe( '2016-12-31T17:00:00.000Z' );
    expect( changeMode ).toHaveBeenCalledTimes( 1 );
    expect( changeMode.mock.calls[ 0 ][ 0 ] ).toBe( 'months' );

    meta.component.container.querySelector( 'thead tr th:nth-child(2)' ).click();
    expect( changeMode ).toHaveBeenCalledTimes( 2 );
    expect( changeMode.mock.calls[ 1 ][ 0 ] ).toBe( 'yearRange' );
} );
