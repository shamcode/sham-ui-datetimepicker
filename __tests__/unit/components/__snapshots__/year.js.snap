// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`change 1`] = `
Object {
  "Options": Object {
    "ID": "component",
    "change": [Function],
    "classForDate": [Function],
    "clickOnHeader": [Function],
    "date": 2020-04-23T06:02:17.000Z,
    "directives": Object {
      "onclick": [Function],
    },
    "header": [Function],
    "isDateSelectable": [Function],
    "items": [Function],
    "nextDelta": 10,
    "nextSymbol": "▸",
    "prevDelta": -10,
    "prevSymbol": "◂",
    "select": [Function],
    "showChange": true,
    "today": 2020-04-25T08:03:30.000Z,
    "value": 2020-04-23T06:02:17.000Z,
  },
  "html": "
<thead>
  <tr>
    <th><span>◂</span></th>
    <th colspan=\\"5\\" class=\\"picker-switch\\">2015 - 2026</th>
    <th><span>▸</span></th>
  </tr>
</thead>
<tbody>
  <tr>
    <td colspan=\\"7\\"><span class=\\"year\\">2015</span><span class=\\"year\\">2016</span><span class=\\"year\\">2017</span><span class=\\"year\\">2018</span><span class=\\"year\\">2019</span><span class=\\"year active\\">2020</span><span class=\\"year disabled\\">2021</span><span class=\\"year disabled\\">2022</span><span class=\\"year disabled\\">2023</span><span class=\\"year disabled\\">2024</span><span class=\\"year disabled\\">2025</span><span class=\\"year disabled\\">2026</span></td>
  </tr>
</tbody>
",
}
`;

exports[`change 2`] = `
Object {
  "Options": Object {
    "ID": "component",
    "change": [Function],
    "classForDate": [Function],
    "clickOnHeader": [Function],
    "date": 2020-04-23T06:02:17.000Z,
    "directives": Object {
      "onclick": [Function],
    },
    "header": [Function],
    "isDateSelectable": [Function],
    "items": [Function],
    "nextDelta": 10,
    "nextSymbol": "▸",
    "prevDelta": -10,
    "prevSymbol": "◂",
    "select": [Function],
    "showChange": true,
    "today": 2020-04-25T08:03:30.000Z,
    "value": 2020-04-23T06:02:17.000Z,
  },
  "html": "
<thead>
  <tr>
    <th><span>◂</span></th>
    <th colspan=\\"5\\" class=\\"picker-switch\\">2015 - 2026</th>
    <th><span>▸</span></th>
  </tr>
</thead>
<tbody>
  <tr>
    <td colspan=\\"7\\"><span class=\\"year\\">2015</span><span class=\\"year\\">2016</span><span class=\\"year\\">2017</span><span class=\\"year\\">2018</span><span class=\\"year\\">2019</span><span class=\\"year active\\">2020</span><span class=\\"year disabled\\">2021</span><span class=\\"year disabled\\">2022</span><span class=\\"year disabled\\">2023</span><span class=\\"year disabled\\">2024</span><span class=\\"year disabled\\">2025</span><span class=\\"year disabled\\">2026</span></td>
  </tr>
</tbody>
",
}
`;

exports[`change 3`] = `
Object {
  "Options": Object {
    "ID": "component",
    "change": [Function],
    "classForDate": [Function],
    "clickOnHeader": [Function],
    "date": 2020-04-23T06:02:17.000Z,
    "directives": Object {
      "onclick": [Function],
    },
    "header": [Function],
    "isDateSelectable": [Function],
    "items": [Function],
    "nextDelta": 10,
    "nextSymbol": "▸",
    "prevDelta": -10,
    "prevSymbol": "◂",
    "select": [Function],
    "showChange": true,
    "today": 2020-04-25T08:03:30.000Z,
    "value": 2020-04-23T06:02:17.000Z,
  },
  "html": "
<thead>
  <tr>
    <th><span>◂</span></th>
    <th colspan=\\"5\\" class=\\"picker-switch\\">2015 - 2026</th>
    <th><span>▸</span></th>
  </tr>
</thead>
<tbody>
  <tr>
    <td colspan=\\"7\\"><span class=\\"year\\">2015</span><span class=\\"year\\">2016</span><span class=\\"year\\">2017</span><span class=\\"year\\">2018</span><span class=\\"year\\">2019</span><span class=\\"year active\\">2020</span><span class=\\"year disabled\\">2021</span><span class=\\"year disabled\\">2022</span><span class=\\"year disabled\\">2023</span><span class=\\"year disabled\\">2024</span><span class=\\"year disabled\\">2025</span><span class=\\"year disabled\\">2026</span></td>
  </tr>
</tbody>
",
}
`;

exports[`renders correctly 1`] = `
Object {
  "Options": Object {
    "ID": "component",
    "change": [Function],
    "classForDate": [Function],
    "clickOnHeader": [Function],
    "date": 2020-04-23T06:02:17.000Z,
    "directives": Object {
      "onclick": [Function],
    },
    "header": [Function],
    "isDateSelectable": [Function],
    "items": [Function],
    "nextDelta": 10,
    "nextSymbol": "▸",
    "prevDelta": -10,
    "prevSymbol": "◂",
    "select": [Function],
    "showChange": true,
    "today": 2020-04-25T08:03:30.000Z,
    "value": 2020-04-23T06:02:17.000Z,
  },
  "html": "
<thead>
  <tr>
    <th><span>◂</span></th>
    <th colspan=\\"5\\" class=\\"picker-switch\\">2015 - 2026</th>
    <th><span>▸</span></th>
  </tr>
</thead>
<tbody>
  <tr>
    <td colspan=\\"7\\"><span class=\\"year\\">2015</span><span class=\\"year\\">2016</span><span class=\\"year\\">2017</span><span class=\\"year\\">2018</span><span class=\\"year\\">2019</span><span class=\\"year active\\">2020</span><span class=\\"year\\">2021</span><span class=\\"year\\">2022</span><span class=\\"year\\">2023</span><span class=\\"year\\">2024</span><span class=\\"year\\">2025</span><span class=\\"year\\">2026</span></td>
  </tr>
</tbody>
",
}
`;
